Engine/GameEngine.cs

```diff
@@ -0,0 +1,410 @@
+            }
+        }
+
+        return new ErrorResult<Player>("No winner yet!");
+    }
+
+    public static Result AllPlayersRequestingTopUp(GameState gameState)
+    {
+        // Check all players are requesting top up
+        return !gameState.Players.All(p => p.RequestingTopUp)
+            ? new ErrorResult("Not all players are requesting top up")
+            : new SuccessResult();
+    }
+
+    public static Result<GameState> TryTopUp(GameState gameState)
+    {
+        // Check all players are requesting top up
+        Result canTopUpResult = CanTopUp(gameState);
+        if (canTopUpResult is IErrorResult canTopUpError)
+        {
+            return new ErrorResult<GameState>(canTopUpError.Message);
+        }
+
+        // Make sure we have cards in our top up pile
+        if (gameState.Players[0].TopUpCards.Count <= 0)
+        {
+            Result<GameState> replenishResult = ReplenishTopUpCards(gameState);
+            if (replenishResult is ErrorResult<GameState> replenishError) return replenishError;
+            gameState = replenishResult.Data;
+        }
+
+        // Move each top up card to their respective center piles
+        for (var i = 0; i < gameState.CenterPiles.Count; i++)
+        {
+            gameState.CenterPiles[i].Add(gameState.Players[i].TopUpCards.Pop());
+        }
+
+        foreach (Player player in gameState.Players)
+        {
+            player.RequestingTopUp = false;
+        }
+
+        return new SuccessResult<GameState>(gameState);
+    }
+
+    private static Result CanTopUp(GameState gameState)
+    {
+        // Check all players are requesting top up
+        Result allPlayersRequestingTopUp = AllPlayersRequestingTopUp(gameState);
+        if (allPlayersRequestingTopUp is ErrorResult allPlayersRequestingTopUpError)
+        {
+            return new ErrorResult(allPlayersRequestingTopUpError.Message);
+        }
+
+        return new SuccessResult();
+    }
+
+    public static Result<GameState> ReplenishTopUpCards(GameState gameState)
+    {
+        // combine middle piles
+        List<Card> combinedCenterPiles = gameState.CenterPiles.SelectMany(cp => cp, (list, card) => card).ToList();
+
+        if (combinedCenterPiles.Count < 1)
+        {
+            return new ErrorResult<GameState>("Can't replenish top up piles without any center cards");
+        }
+
+        // Shuffle them
+        combinedCenterPiles.Shuffle(gameState.Settings?.RandomSeed);
+
+        // Split all but center piles count into top up piles
+        int topUpPileSize = combinedCenterPiles.Count / gameState.Players.Count;
+        foreach (Player player in gameState.Players)
+        {
+            player.TopUpCards = combinedCenterPiles.PopRange(topUpPileSize);
+        }
+
+        // Reset the center piles
+        var newCenterPiles = new List<List<Card>>();
+        for (var i = 0; i < gameState.CenterPiles.Count; i++) newCenterPiles.Add(new List<Card>());
```

Comment:
nitpick: Personally I always use a code style that doesn't allow one line statements without `{ }` after controle blocks `if`, `for`, `while` etc.

On another note it might be worth looking at `https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.repeat?view=net-6.0` which would allow you to do this line and make it more neat and readable in my opinion